### PR TITLE
fix(Errors): Enable nativeSerializer to handle Error objects

### LINKED ISSUE #239: Support for Error class in NativeSerializer (feature request)
### Description
Consider bugsnag code like below that tries to leverage metadata to track details of error.

```:JavaScript
const err = new Error('this is cause');
const params = { foo: 1 };
bugsnag.notify(new Error('This is an example'), (report) => {
  report.metadata.account = { cause: err, params };
});
```

It will create meatadata like:
```
{
  cause: {},
  params: { foo: 1 }
}
```

Note cause is logged as empty object.

**Describe the solution you'd like**
Implement handling for Error type within https://github.com/bugsnag/bugsnag-react-native/blob/master/src/NativeSerializer.js so it'll log relevant info (eg., error message)

### PR DESCRIPTION
**Pull request by Cawllec**
## Goal

Fixes #239, extracting the stack and message from a given `Error` class before serialising it.

## Design

Error objects do not have an iterator therefore are ignored by the `for` loop.  My initial thoughts were to add an iterator using `Object.keys` to extract all properties into a map, but the Error object does not respond to this either.

### Added

Added non-exported method `extractErrorDetails` to `NativeSerializer` that takes an error and returns an object with the Errors `message` and `stack` properties.

## Tests

Added a unit test to `NativeSerializer.test`

### Linked issues

Fixes #239 

For the submitter, initial self-review:

- [ ] Commented on code changes inline explain the reasoning behind the approach
- [ ] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [ ] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [ ] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [x] Consistency across platforms for structures or concepts added or modified
- [x] Consistency between the changeset and the goal stated above
- [x] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [x] Usage friction - is the proposed change in usage cumbersome or complicated?
- [x] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [x] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [x] Thoroughness of added tests and any missing edge cases
- [x] Idiomatic use of the language

### CODE REVIEW

File: src/NativeSerializer.js [hunk 1]
```diff
@@ -7,6 +7,9 @@ const allowedMapObjectTypes = [ 'string', 'number', 'boolean' ]
 const serializeForNativeLayer = (map, maxDepth = 10, depth = 0, seen = new Set()) => {
   seen.add(map)
   const output = {}
+  if (map instanceof Error) {
```
Comment (bengourley) [inline src/NativeSerializer.js]: I would use the [iserror](https://github.com/yefremov/iserror)  module for more robust `Error` detection. `instanceof` checks alone can yield [unexpected results](https://groups.google.com/forum/#!msg/comp.lang.javascript/XTWYCOwC96I/70rNoQ3L-xoJ). 

We make similar use of this module in [bugsnag-js](https://github.com/bugsnag/bugsnag-js/blob/9b2309d2a3adedb187dd908e7c4850d228ea2c46/browser/plugins/unhandled-rejection.js#L74).

File: src/NativeSerializer.js [hunk 2]
```diff
@@ -35,4 +38,12 @@ const serializeForNativeLayer = (map, maxDepth = 10, depth = 0, seen = new Set()
   return output
 }
 
+const extractErrorDetails = (err) => {
+  const {message, stack} = err
```
Comment (bengourley) [inline src/NativeSerializer.js]: style nit that the linter doesn't pick up on:

```js
const { message, stack } = err
```

File: src/NativeSerializer.js [hunk 3]
```diff
@@ -35,4 +38,12 @@ const serializeForNativeLayer = (map, maxDepth = 10, depth = 0, seen = new Set()
...
+  return {
+    message: message,
```
Comment (bengourley) [inline src/NativeSerializer.js]: there's nice new-ish js shortand for this:

```js
return { message, stack }
```

but you should also return the `name` and potentially the `code` property (this is only relevant to DOMexceptions, can you get them in RN? 🤔):

Comment (Cawllec) [reply src/NativeSerializer.js]: From what I've been reading the `message` and `stack` are the only guarantees.  I don't want to be dumping a whole load of `undefined`s into the metadata, so left out everything else.

I could add test whether properties are assigned and extract them if so?

File: src/__tests__/NativeSerializer.test.js [hunk 4]
```diff
@@ -134,3 +134,14 @@ test('serializeForNativeLayer handles deep nesting', () => {
...
+test('serializeForNativeLayer handles error objects', () => {
+  const err = new Error('Oh no')
+  const serialized = serializeForNativeLayer(err)
+  expect(serialized['message']).toEqual({
+    type: 'string',
+    value: 'Oh no'
+  })
+  expect(serialized['stack']['type']).toEqual('string')
+  expect(serialized['stack']['value'].length).toBeGreaterThan(0)
+})
```
Comment (bengourley) [inline src/__tests__/NativeSerializer.test.js]: add an assertion for `name` property, and if the `code` property is added, assertions for the inclusion/exclusion of that.

**Comment by Cawllec** [comment]
Made changes based on the above review.  @fractalwrench, is there a way I can test that the `iserror` module is in-fact bundled with the module?

**Comment by kattrali** [comment]
@Cawllec building the package via `yarn pack`/`npm pack` and then installing it into a sample app (something like `yarn add /path/to/bugsnag-react-native-2.9.4.tgz`) should both:

* Show that `iserror` is installed in the log output
* Be immediately testable

**Comment by Cawllec** [comment]
After sorting out my local examples I can confirm the `iserror` module is being installed.  The errors attached to metadata are also coming through as expected when manually testing.
